compiler/rustc_transmute/src/maybe_transmutable/mod.rs

```diff
@@ -75,14 +76,14 @@ mod rustc {
                 let dst = Tree::from_ty(dst, context);
 
                 match (src, dst) {
-                    // Answer `Yes` here, because 'unknown layout' and type errors will already
+                    // Answer `Ok(None)` here, because 'unknown layout' and type errors will already
                     // be reported by rustc. No need to spam the user with more errors.
-                    (Err(Err::TypeError(_)), _) => Err(Answer::Yes),
-                    (_, Err(Err::TypeError(_))) => Err(Answer::Yes),
-                    (Err(Err::Unknown), _) => Err(Answer::Yes),
-                    (_, Err(Err::Unknown)) => Err(Answer::Yes),
-                    (Err(Err::Unspecified), _) => Err(Answer::No(Reason::SrcIsUnspecified)),
-                    (_, Err(Err::Unspecified)) => Err(Answer::No(Reason::DstIsUnspecified)),
+                    (Err(Err::TypeError(_)), _)
+                    | (_, Err(Err::TypeError(_)))
+                    | (Err(Err::Unknown), _)
+                    | (_, Err(Err::Unknown)) => Err(Ok(None)),
```

Comment:
I don't think it's correct to treat `Unknown` like this. We will eventually reach this when we support transmuting `&T` that references a generic parameter `T`, since if `T` is not sized then it'll have an unknown layout, and I don't think we want to treat that as transmutability?

Side-note, using `Result` as if it were `Either` here is confusing. Maybe it should be lifted into its own variant. Same with the `Result` inside of that `Result`, which is (I think) `Answer<T>` -- why not just keep that as its own enum with different variants corresponding to "`Yes`", "`OnlyIf`", and "`No`"?

Side-note of a side-note: `map_layouts` seems like a really obtuse way of unpacking a `MaybeTransmutableQuery`. We really shouldn't be doing all this, tbh.

Reply:
So to summarize, maybe let's get rid of `map_layouts` altogether?

Reply:
> I don't think it's correct to treat `Unknown` like this. We will eventually reach this when we support transmuting `&T` that references a generic parameter `T`, since if `T` is not sized then it'll have an unknown layout, and I don't think we want to treat that as transmutability?

The purpose of returning `Ok(None)` (previously `Answer::Yes`) is to just exit early and not emit any errors. This is because an error will already be emitted by rustc because of the unknown layout. It's not that we are saying "yes, transmutation is possible", but rather we just quit out here and let rustc emit some other error.

> 
> Side-note, using `Result` as if it were `Either` here is confusing. Maybe it should be lifted into its own variant. Same with the `Result` inside of that `Result`, which is (I think) `Answer<T>` -- why not just keep that as its own enum with different variants corresponding to "`Yes`", "`OnlyIf`", and "`No`"?

Yeah, I agree the nested results is confusing to read. I'll  see if I can improve it.

> 
> Side-note of a side-note: `map_layouts` seems like a really obtuse way of unpacking a `MaybeTransmutableQuery`. We really shouldn't be doing all this, tbh.

For sure, it's not easy to read/understand. Let's refactor it in a separate PR, I think it will add a bunch more changes to do it in here.

Reply:
> The purpose of returning Ok(None) (previously Answer::Yes) is to just exit early and not emit any errors. This is because an error will already be emitted by rustc because of the unknown layout.

Where is that error emitted? Rust will not, in general, emit an error on an *unknown* layout. Maybe I'm misunderstanding something.

Reply:
For example, this code relies on layout computation returning "unknown" not emitting an error:

```
#![feature(pointer_like_trait)]

use std::ops::Deref;
use std::marker::PointerLike;

impl<T: PointerLike> Wrapper<T> {
    fn do_it(self) {
        println!("is pointer-like");
    }
}

impl WrapperDeref {
    fn do_it(&self) {
        println!("is not pointer-like");
    }
}

struct Wrapper<T>(T);

impl<T> Deref for Wrapper<T> {
    type Target = WrapperDeref;
    
    fn deref(&self) -> &WrapperDeref { &WrapperDeref }
}

struct WrapperDeref;

fn test<T>(t: T) {
    Wrapper(&t).do_it();
    Wrapper(t).do_it();
}

fn main() {
    test(1i32);
}
```

Inside of the method `test`, we first check if `T: PointerLike` holds, which is going to return `LayoutError::Unknown`, which is *not* a fatal error.

Reply:
Hmm okay thanks for the detailed comment, I guess I should change the `Unknown` case to emit an error then...

@jswrenn do you remember why you had left the comment on line 79 saying that `Unknown` would already be reported?

Reply:
It's not *currently* reachable because we don't even check transmutability for things that need substitution (which I assume could get lifted). 

But I assume it will, and I'm not confident that we'll remember to set this to fail the predicate, and it'll just cause the transmutability predicate to succeed.